src/send/mod.rs

```diff
@@ -87,8 +96,34 @@ fn fail() -> data::Outcome<Email, ValidationError>
   ))
 }
 
-#[post("/send", format = "application/json", data = "<_email>")]
-fn handler(_email: Email) -> Json<Value>
+#[post("/send", format = "application/json", data = "<email>")]
+fn handler(email: Email) -> Result<Json<Value>, Error>
 {
-  Json(json!({ "messageId": 0 }))
+  let cc = if let Some(cc) = email.cc {
+    cc.clone()
+  } else {
+    Vec::new()
+  };
+  let html = if let Some(html) = email.body.html {
+    Some(html.clone())
+  } else {
+    None
+  };
+  let provider = if let Some(provider) = email.provider {
+    Some(provider.clone())
+  } else {
+    None
+  };
+
+  match PROVIDERS.send(
+    email.to.clone(),
+    cc,
+    email.subject.clone(),
+    email.body.text.clone(),
```

Comment:
Naively, I would expect all of these strings to say alive for the entire duration of the request, and hence to be able to pass borrows rather than clones.  For my own learning purposes, could you please explain why the clones are needed here?

Reply:
There were a few steps that led me to this point so I'll recount them in turn as maybe I took a wrong step along the way:

1. [Rusoto's `Content` type](https://rusoto.github.io/rusoto/rusoto_ses/struct.Content.html) takes allocated `String` instances, so at some point I was going to have to choose between cloning the strings or passing ownership to Rusoto.

2. Passing ownership to Rusoto would preclude us from using the strings after it's been invoked. Maybe we want them afterwards for logging or some other post-send activity like setting up bounce handlers.

3. At that point the choice is between either cloning inside the SES provider and passing it `&str` or cloning here and passing `String`. To some extent the right decision here depends on what the implementation will look like for other providers. If others also need cloned strings, it makes sense to do that once, outside. If not it makes sense for that to be SES-specific behaviour and pass string slices. I thought about it for a bit, wasn't sure which would be the case and took a punt on cloning outside.

I'm happy to change this to pass ownership of the strings to Rusoto, or to clone them inside the SES provider, whichever you're more comfortable with. I wasn't really sure which would turn out to be right in the long run and optimistically hoped it would all come out in the wash. :smile:

Reply:
> I'm happy to change this to pass ownership of the strings to Rusoto, or to clone them inside
> the SES provider, whichever you're more comfortable with.

I'm very happy to leave them as `clone`s here, just trying to learn as I go :-)

IIUC this is a pretty common evolution for rust code, start off with a bunch of clones until you stabilize the interefaces and figure out which can be borrows etc, so 👍 let's not block on getting that all exactly right.

Reply:
Fwiw, after sleeping on this I decided to switch to slices in the `Provider` interface and then do the cloning inside the `Ses` provider as a specific step.